pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,348 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServices(vServices []Config) (out []Config) {
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[vs.Namespace+vs.Name] = vs
+			continue
+		}
+
+		// root vs
+		if isRootVs(rule) {
+			rootVses = append(rootVses, vs)
+			continue
+		}
+
+		// the others are normal vs without delegate
+		out = append(out, vs)
+	}
+
+	// 2. merge delegates and root
+	for _, root := range rootVses {
+		rootVs := root.Spec.(*networking.VirtualService)
+		notFound := false
+		mergedRoutes := []*networking.HTTPRoute{}
+		for _, route := range rootVs.Http {
+			// it is root vs with delegate
+			if route.Delegate != nil {
+				delegate, ok := delegatesMap[route.Delegate.Namespace+route.Delegate.Name]
+				if !ok {
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore the root as well
+					notFound = true
+					break
+				}
+				vs := delegate.Spec.(*networking.VirtualService)
+				merged := mergeHttpRoute(route, vs.Http)
+				mergedRoutes = append(mergedRoutes, merged...)
+			} else {
+				mergedRoutes = append(mergedRoutes, route)
+			}
+		}
+		if !notFound {
+			rootVs.Http = mergedRoutes
+			out = append(out, root)
+		}
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHttpRoute(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		// suppose there are N1 match conditions in root, N2 match conditions in delegate
+		// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
+		merged, conflict := mergeHTTPMatchRequests(root.Match, subRoute.Match)
```

Comment:
We should have an `istioctl analyze`r for this?

Reply:
make sense

Reply:
I would add a TODO. Once this API is approved, i will add it